### DISCUSSION TITLE
Can I change "s:" to "soap:" in response?

### DISCUSSION THREAD

**Discussion by michael-kopt** [CoreWCF/CoreWCF#1110] [General]
Hi,

I'm moving a legacy project from WCF to CoreWCF, everything works fine, but for some reason for other services that are getting responses from CoreWCF is important to have "soap:" in response. Is it possible?

Thanks in advance.

**Comment by kjell-walvag** [comment]
Has the same issue. We do not have control of the client implementations. 
So "s:" prefix should be the same as the old service = "soap:"
How do we change "s:" to "soap:"?
Anyone?

**Reply by mconnew** [reply]
In xml, the prefix (in this case s or soap) is defined by a namespace definition. Here's an example XML snippet to demonstrate:
```xml
<a:MyElement xmlns:a="http://example.org/myelements.xsd">Hello</a:MyElement>
```
In this snippet, there's a namespace definition via the attribute `xmlns:prefix` which defines a namespace prefix to namespace mapping. That prefix is then valid within the scope of that element, including any child XML elements. If `MyElement` didn't just contain a string but had other XML elements, you could use the `a:` prefix in any of the contained XML. You can't use the prefix for an element after the closing tag as the scope has been closed and it's not longer a valid namespace prefix. The actual prefix used doesn't matter as long as it's a valid XML Name (no special characters etc). The following XML snippet is semantically equivalent to the earlier one and should be parsed identically by any consuming code.
```xml
<potato:MyElement xmlns:potato="http://example.org/myelements.xsd">Hello</potato:MyElement>
```
It's a key value pair for whenever a parser wants to know what uri namespace an entity is in (could be an element, or an attribute), but the actual prefix key doesn't matter as it's defined in the document itself. So the following 2 XML snippets are equivalent:
```xml
<soap:Envelope
xmlns:soap="http://www.w3.org/2003/05/soap-envelope/"
soap:encodingStyle="http://www.w3.org/2003/05/soap-encoding">
.... SOAP headers and body go here ...
</soap:Envelope>
```

```xml
<s:Envelope
xmlns:s="http://www.w3.org/2003/05/soap-envelope/"
s:encodingStyle="http://www.w3.org/2003/05/soap-encoding">
.... SOAP headers and body go here ...
</s:Envelope>
```

I have seen one SOAP stack which hard coded in what they expect the prefix to be, but that's a buggy implementation as it should be ambivalent to which prefix is used, as long as it's valid XML.  

There is one limitation I know of in WCF/CoreWCF which there isn't an easy way to fix where any namespaces used unique to a SOAP header, or in the SOAP body needs to be defined in the headers or the body element as a new instance of an XmlReader is used and the parent namespaces aren't propagated in, even though strictly speaking they should be. I've only seen that be an issue once in over a decade of working on WCF.

**Reply by mconnew** [reply]
I just realized I didn't actually answer your question. I'm going to dig into the code to see where we define what we're going to use as the prefix and work out if it can be easily modified. If not with current code, I might be able to add an API to modify it.

**Reply by mconnew** [reply]
I found where it's being written. You can change it, but it's not super easy. The prefix is written by the CoreWCF.Channels.Message class and is used in 3 places. [OnWriteStartBody](https://github.com/CoreWCF/CoreWCF/blob/375ae8a8a5a5a178852ed8f4eb3a276ebbcf1155/src/CoreWCF.Primitives/src/CoreWCF/Channels/Message.cs#L467), [OnWriteStartEnvelope](https://github.com/CoreWCF/CoreWCF/blob/375ae8a8a5a5a178852ed8f4eb3a276ebbcf1155/src/CoreWCF.Primitives/src/CoreWCF/Channels/Message.cs#L527), and [OnWriteStartHeaders](https://github.com/CoreWCF/CoreWCF/blob/375ae8a8a5a5a178852ed8f4eb3a276ebbcf1155/src/CoreWCF.Primitives/src/CoreWCF/Channels/Message.cs#L538).  

You will need to create an IDispatchMessageInspector which replaces the message (it's a ref parameter so you can completely replace it) with your own implementation of Message. Your implementation will derive from Message, and wrap the instance which was passed in to the message inspector. You will need to override those 3 methods replacing the implementation with one which uses the namespace prefix that you want. I'm not sure what's going to be the easiest way to get everything else written out as Message has a bunch of state information it maintains so you can't just pass through all calls. This is quite a bit of work as there's likely a lot of implementation you will need to copy across to get everything to work.  

A less performant but probably a lot easier to write implementation would be to create a custom Message class and override the WriteMessageAsync method.  Create a MemoryStream and wrap it in an XmlDictionaryWriter, then pass that to the original message WriteMessageAsync method. Then get a string from the MemoryStream, fix up the prefix in the message, then create an XmlDictionaryReader from that. On the passed in XmlDictionaryWriter,  call XmlDictionaryWriter.WriteNode using this reader, This is relatively allocation heavy and isn't the faster way, but if your service doesn't have very big responses, it shouldn't be too bad. I would definitely validate the performance hit of doing this.  

Ultimately the problem is the receiving client isn't conforming to the XML specification and is hard requiring a specific namespace. The purest in me says that library needs to fix their implementation, but the pragmatic solution is to make the change in CoreWCF. If this issue gets a lot of attention, I might add an to enable changing the prefix for the namespaces. I think adding a Prefix property to EnvelopeVersion might be the most elegant way. Then it would be a property of the binding (on the MessageEncoder binding element).

**Comment by kjell-walvag** [comment]
Thank you for your answer. For the moment we will not upgade to .Net Core. Too many risk with the external clients (Some are old COBOL programs! where the programmers are retired)
Hope in future version a Prefix property or something is added.

**Reply by mconnew** [reply]
What are you using currently? This behavior is the same between WCF on .NET Framework and CoreWCF

**Comment by kjell-walvag** [comment]
Hi

We are using .Net Framework 4.8 in the current solution. The current SOAP service is created with Visual Studio.

The input-outputs:

```
.Net Framework 4.8
-------------------------------
<Envelope xmlns="http://schemas.xmlsoap.org/soap/envelope/">
    <Body>
        <Ping xmlns="https://autocollect.visma.com/">
            <message>xxx</message>
        </Ping>
    </Body>
</Envelope>

<?xml version="1.0" encoding="utf-8"?>
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
    <soap:Body>
        <PingResponse xmlns="https://autocollect.visma.com/">
            <PingResult>xxx</PingResult>
        </PingResponse>
    </soap:Body>
</soap:Envelope>



CoreWCF
-------------------------------
<Envelope xmlns="http://schemas.xmlsoap.org/soap/envelope/">
    <Body>
        <Ping xmlns="https://autocollect.visma.com/">
            <message>xxx</message>
        </Ping>
    </Body>
</Envelope>

<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/">
    <s:Body>
        <PingResponse xmlns="https://autocollect.visma.com/">
            <PingResult>xxx</PingResult>
        </PingResponse>
    </s:Body>
</s:Envelope>
```

**Reply by mconnew** [reply]
Is your current service using asmx Web Services on .NET Framework? The code for emitting the Envelope and Body is identical between WCF and CoreWCF so there shouldn't be any difference there.

**Reply by kjell-walvag** [reply]
I have created a new WebService in VisualStudio 2022 using project template for .Net 4.8 Web-service. The created service is using "soap" as namespace prefix shown in the example above.
If creating a new Web-service in Visual Studio 2022 using project template for CoreWCF (downloaded and added) the namespace prefix is "s"

**Reply by kjell-walvag** [reply]
Yes a asmx web service

**Reply by mconnew** [reply]
I found you a solution, someone already solved it [in this blog post](https://www.vanacosmin.ro/Articles/Read/WCFEnvelopeNamespacePrefix). The link to a microsoft blog post with a different approach has moved, the [current link is here](https://learn.microsoft.com/en-us/archive/blogs/carlosfigueira/changing-prefixes-in-xml-responses-for-wcf-services).  

The vanacosmin blog post will be as fast as without the fix so you shouldn't see any performance issues. I believe everything you need is in CoreWCF, except some methods have changed to async. For each method which there's an overridden implementation, check if there's an async implementation and override that. Some have the old sync and the new async for compat reasons, you should just have to override the async one only.

**Reply by kjell-walvag** [reply]
Thanks for your help Matt! [Solution ](https://www.vanacosmin.ro/Articles/Read/WCFEnvelopeNamespacePrefix) works!
Had to change one line in the code from:
var serializerBehavior = operationDescription.Behaviors.Find<DataContractSerializerOperationBehavior>();
to:
 var serializerBehavior = (
                from o in operationDescription.OperationBehaviors
                where o.GetType().Name.ToLower() == "datacontractserializeroperationbehavior"
                select o).FirstOrDefault();

**Reply by mconnew** [reply]
I'm glad you have a solution. And the good news is there shouldn't be any performance overhead with this approach.

**Reply by danreg** [reply]
I wanted to add a solution without using Attributes. You can put this code to your Configure method in Startup

` serviceBuilder.AddServiceEndpoint<InterfaceService, IInterfaceService>(new BasicHttpBinding(), "/int/interfaceservice.asmx", (c) =>
                {
                    CustomFormaterBehavior customFormaterBehavior = new CustomFormaterBehavior();
                    c.Contract.Operations.ForEach(description => { description.OperationBehaviors.Add(customFormaterBehavior); });
                });`


And I used this approach instead of the example @kjell-walvag gave
`            IOperationBehavior? serializerBehavior = operationDescription.OperationBehaviors.FirstOrDefault(x => x is DataContractSerializerOperationBehavior);
`
